### PR TITLE
opal/sync: fix race condition

### LINKED ISSUE #1813: Overlap test sometimes failing in Mellanox Jenkins with Yalla
Mellanox Jenkins is showing a periodic problem with the overlap test using Yalla.  There has been some discussion on https://github.com/open-mpi/ompi-release/pull/1240 about what the actual root cause is, starting around here: https://github.com/open-mpi/ompi-release/pull/1240#issuecomment-227990044.

Please continue the discussion here: @jladd-mlnx @miked-mellanox @artpol84 @nysal @hjelmn 

We need to decide if this is a v2.0.0 or v2.0.1 issue.  What is the scope of the problem, and how quickly can it be fixed?  I'm tentatively marking this as v2.0.0, just so that it comes up on our daily discussions.

I'm opening a separate issue to track the discussion that was occurring on https://github.com/open-mpi/ompi-release/pull/1240 (so that 1240 can go back to its regularly scheduled programming).

### PR DESCRIPTION
**Pull request by hjelmn**
This commit fixes a race condition discovered by @artpol84. The race
happens when a signalling thread decrements the sync count to 0 then
goes to sleep. If the waiting thread runs and detects the count == 0
before going to sleep on the condition variable it will destroy the
condition variable while the signalling thread is potentially still
processing the completion. The fix is to add a non-atomic member to
the sync structure that indicates another process is handling
completion. Since the member will only be set to false by the
initiating thread and the completing thread the variable does not need
to be protected. When destoying a condition variable the waiting
thread needs to wait until the singalling thread is finished.

Thanks to @artpol84 for tracking this down.

Fixes #1813

Signed-off-by: Nathan Hjelm <hjelmn@lanl.gov>

### CODE REVIEW

**Comment by hjelmn** [comment]
@artpol84 Getting this running through Jenkins. Will try to get it retested multiple times to see if it fails. In theory it shouldn't.

**Comment by jladd-mlnx** [comment]
bot:retest

**Comment by hjelmn** [comment]
:bot:retest:

**Comment by hjelmn** [comment]
:bot:retest:

**Comment by jladd-mlnx** [comment]
Retest after adding the original Jenkins command line with ob1 (allow UCX to load and unload.) 

bot:retest

**Comment by hjelmn** [comment]
I can make this even faster as the lock/unlock is not really needed for the condition signal here. Will leave that for later though.

**Comment by hjelmn** [comment]
Looking good. Sending through again.

:bot:retest:

**Comment by artpol84** [comment]
@hjelmn I guess that race condition is still possible here
1. `(sync)->signalling` is initialized to `false` and this is correct because you may wait late (after all `wait_sync_update` was called)
2, Assume instruction reordering in `wait_sync_update` (correct be if I'm wrong, but it's possible):

| count | PML progress  | MPI_Wait thread |
| ------------- | ------------- | ------------- |
| 1 | OPAL_THREAD_ADD32(&sync->count, -1)  |   |
|0 |  | if( sunc->count > 0 ) - false => exit the loop |
| 0 | | running till the end and free'ing `sync` |
|0| WAIT_SYNC_SIGNAL(sync); - error | |

Same error here. I also assume there will be problems with several independently running PML progress threads. 

So it seems that you need a memory fence here:

```C
static inline void wait_sync_update(ompi_wait_sync_t *sync, int updates, int status)
  {
     sync->signalling = true;

     mem_fence(); // <---------   

     if( OPAL_LIKELY(OPAL_SUCCESS == status) ) {
          if( 0 != (OPAL_THREAD_ADD32(&sync->count, -updates)) ) {
              return;
         }
     } else {
         /* this is an error path so just use the atomic */
         opal_atomic_swap_32 (&sync->count, 0);
         sync->status = OPAL_ERROR;
     }
     WAIT_SYNC_SIGNAL(sync);
 }
```

This race will be much more rare compared to original one, and harder to track.

**Comment by hjelmn** [comment]
Instruction reordering will not happen. pthread calls are effectively a memory barrier. We could add an isync just to be safe if you want think it is warranted. We have isync for both ppc and arm.

**Comment by hjelmn** [comment]
Oh, wait. *before* the atomic. Sorry, didn't catch that part. You are absolutely correct.

**Comment by hjelmn** [comment]
An atomic wmb would be fine there. An isync as I suggestedmight be too weak.

**Comment by artpol84** [comment]
Setting of signal and changing count doesn't have pthreads, they are coming
lately

суббота, 25 июня 2016 г. пользователь Nathan Hjelm написал:

> Instruction reordering will not happen. pthread calls are effectively a
> memory barrier. We could add an isync just to be safe if you want think it
> is warranted. We have isync for both ppc and arm.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/open-mpi/ompi/pull/1816#issuecomment-228507294>, or mute
> the thread
> <https://github.com/notifications/unsubscribe/AHL5Pkr67OoyDmNxE209ANgTPm1ZPMfpks5qPKI0gaJpZM4I-HOt>
> .
>


-- 
-----
Best regards, Artem Polyakov
(Mobile mail)

**Comment by hjelmn** [comment]
@artpol84 Yup. Just caught that after hitting comment :)

**Comment by hjelmn** [comment]
Will add the barrier now.

**Comment by hjelmn** [comment]
And correct to the more common signaling spelling :D

**Comment by hjelmn** [comment]
@artpol84 Thanks for catching that. Should be fixed now.

**Comment by hjelmn** [comment]
Won't commit this until I get a +1 from @artpol84.

Looks like we had a finalize hang. Looks unrelated. @artpol84 Any clue as to what happened?

**Comment by hjelmn** [comment]
rdmacm failure here too. I will merge the rdmacm fix I have open and see if it fixes it.

**Comment by hjelmn** [comment]
:bot:retest:

**Comment by hjelmn** [comment]
@artpol84 Feel free to open a PR vs my branch and add the mellanox copyright. I prefer the commit adding a copyright line be from a member the copyrighting organization.

File: opal/threads/wait_sync.h [hunk 1]
```diff
@@ -75,6 +89,9 @@ static inline int sync_wait_st (ompi_wait_sync_t *sync)
  */
 static inline void wait_sync_update(ompi_wait_sync_t *sync, int updates, int status)
 {
+    sync->signaling = true;
+    /* ensure the signalling value is commited before updating the count */
+    opal_atomic_wmb ();
```
Comment (artpol84) [inline opal/threads/wait_sync.h]: @hjelmn I may be wrong but it seems that wmb addresses CPU-level reordering possibility.
Do we address compiler reordering?

Comment (artpol84) [inline opal/threads/wait_sync.h]: I mean adding something like `asm volatile("" ::: "memory");` may be needed

Comment (artpol84) [inline opal/threads/wait_sync.h]: Ok, this is already done in `wmb()`

File: opal/threads/wait_sync.h [hunk 2]
```diff
@@ -75,6 +89,9 @@ static inline int sync_wait_st (ompi_wait_sync_t *sync)
...
     if( OPAL_LIKELY(OPAL_SUCCESS == status) ) {
         if( 0 != (OPAL_THREAD_ADD32(&sync->count, -updates)) ) {
```
Comment (artpol84) [inline opal/threads/wait_sync.h]: @hjelmn @jladd-mlnx I have a question here.
Consider we have 2 concurrently running wait_sync_updates. Opal thread add is atomic in the sense that intermediate results of the op will not intermix.
**however** we can get both of them receiving 0 if they called this primitive together.
This means another possible race condition:
* Both threads will go to release path
* First thread will release signaling flag
* waiting thread will release sync
* second thread running wait_sync_updates will hit the same bug.

Comment (artpol84) [inline opal/threads/wait_sync.h]: By the "same bug" I mean signaling on uninitialized condition.

Comment (artpol84) [inline opal/threads/wait_sync.h]: I guess I'm not correct here.
Only one (the last) thread will get 0.
So no race here.

Comment (bosilca) [inline opal/threads/wait_sync.h]: Nice catch. The scope of this race condition is restricted to cases where
the synchronization primitive is used in a wait or waitall but not waitany
or waitsome (where the sync primitive is atomically removed from the
requests before release).

As the only thing we need is to guarantee that the signaling thread is out
of the signaling sequence, so we can start with the signaling field set to
true. This will save the memory barrier in the wait_sync_update, and will
only require the change in WAIT_SYNC_SIGNAL.

Also the error path is now broken, the following patch is necessary.
```
diff --git a/opal/threads/wait_sync.h b/opal/threads/wait_sync.h
index b29f01c..d2b5003 100644
--- a/opal/threads/wait_sync.h
+++ b/opal/threads/wait_sync.h
@@ -80,9 +80,10 @@ static inline void wait_sync_update(ompi_wait_sync_t
*sync, int updates, int sta
             return;
         }
     } else {
-        /* this is an error path so just use the atomic */
-        opal_atomic_swap_32 (&sync->count, 0);
+        /* We need the status to be visible when the count become 0 */
         sync->status = OPAL_ERROR;
+        opal_atomic_wmb();
+        opal_atomic_swap_32 (&sync->count, 0);
     }
     WAIT_SYNC_SIGNAL(sync);
 }
```

Comment (artpol84) [inline opal/threads/wait_sync.h]: @bosilca I'll add this as a separate commit in my PR https://github.com/open-mpi/ompi/pull/1815 on your behalf.

**Comment by artpol84** [comment]
Looks like waitany and waitsome are broken. I filed the PR to @hjelmn branch:
https://github.com/artpol84/ompi/commit/49b1c0ffadde8341cbf1a0186325bd3614f4634a

Can be easily reproduced with following modifications of overlap test:
[overlap_waitany.txt](https://github.com/open-mpi/ompi/files/334573/overlap_waitany.txt)
[overlap_waitsome.txt](https://github.com/open-mpi/ompi/files/334574/overlap_waitsome.txt)

@hjelmn @bosilca please, check.

**Comment by artpol84** [comment]
@jladd-mlnx @miked-mellanox I think we (Mellanox) will need to include this tests to our jenkins/MTT suite.
They exercise both corner cases: late completion and early completion of all requests.
We can also add versions without delay to exercise other possible paths.
Since those are race conditions - probably multiple executions is needed.

**Comment by jsquyres** [comment]
@artpol84 It would be good if these could be standalone tests added to the ibm test suite in ompi-tests; then they'll be run in MTT every night by a bunch of organizations.

**Comment by artpol84** [comment]
@jsquyres, sure, we can do that as well.

**Comment by hjelmn** [comment]
@bosilca, @artpol84 Are we all ok with this change the way it is? @artpol84 You can close your PR. Due to a rebase it would no longer merge so I cherry-picked the commit.

**Comment by artpol84** [comment]
@hjelmn, what do you mean? I don't see the wait some waitany fix in this Pr.

**Comment by bosilca** [comment]
I think this patch fixes the original issue, and as such it is ready to go. The performance issues discovered meanwhile should be fixed by a subsequent ticket, once we have the correct fix.

**Comment by artpol84** [comment]
There was a) a bug and b) performance issue.
I PRed a bug fix for wait some and waitany here.
And perf is pending on another request

вторник, 28 июня 2016 г. пользователь bosilca написал:

> I think this patch fixes the original issue, and as such it is ready to
> go. The performance issues discovered meanwhile should be fixed by a
> subsequent ticket, once we have the correct fix.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/open-mpi/ompi/pull/1816#issuecomment-228907225>, or mute
> the thread
> <https://github.com/notifications/unsubscribe/AHL5PoIxz2ev2r02WH_ZDSnOaj2igRMEks5qQF3ogaJpZM4I-HOt>
> .
>


-- 
-----
Best regards, Artem Polyakov
(Mobile mail)

**Comment by artpol84** [comment]
https://github.com/hjelmn/ompi/pull/9/commits/42e6251a25e4b9064f72352a150af1105098899f
This is a bug fix

**Comment by artpol84** [comment]
Without this bugfix the following tests (already mentioned above) are hanging 100%
[overlap_waitany.txt](https://github.com/open-mpi/ompi/files/336056/overlap_waitany.txt)
[overlap_waitsome.txt](https://github.com/open-mpi/ompi/files/336057/overlap_waitsome.txt)

I think PR must fix without new hang introduction.

**Comment by bosilca** [comment]
42e6251 introduces its own issues (I commented on it on the other issue about the possible race condition). A possible fix is to generate a sync_update (which will take care of the signaling field) in the  wait* operation if there is nothing to wait for.

**Comment by artpol84** [comment]
@bosilca I think that your comment https://github.com/open-mpi/ompi/pull/1820#issuecomment-228905750 is reasonable for MPI_Waitany only. So in my understanding it is related to this PR because I want this commit https://github.com/open-mpi/ompi/pull/1820/commits/42e6251a25e4b9064f72352a150af1105098899f to go along with this PR.
I will fix that now. For MPI_Waitsome I believe we are fine.

**Comment by artpol84** [comment]
@hjelmn I rebased my PR https://github.com/hjelmn/ompi/pull/9 and added the fix for the problem that @bosilca pointed, please don't rebase untill we agreed here.

@bosilca please have a look.

**Comment by artpol84** [comment]
@bosilca btw I solved the problem with Waitany with sync_sets/unsets instead of sync_update. I can do the sync_update but with sync_sets/unsets we avoiding one atomic. What do you think?

**Comment by artpol84** [comment]
ohh, and according to the backtrace I had when I was debugging:
https://github.com/open-mpi/ompi-release/pull/1240#issuecomment-227990044

sync_wait will then try to call condition which has internal lock (see the backtrace of the second thread).
So I guess we wont to avoid sync_wait call when possible.

**Comment by artpol84** [comment]
@bosilca @hjelmn I'll be on today's OMPI call available for discussion.

**Comment by bosilca** [comment]
I wont be able to join the call today.

**Comment by artpol84** [comment]
I think I've got all your points now. Will update PRs later today. Thank
you!

вторник, 28 июня 2016 г. пользователь bosilca написал:

> I wont be able to join the call today.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/open-mpi/ompi/pull/1816#issuecomment-229005736>, or mute
> the thread
> <https://github.com/notifications/unsubscribe/AHL5PqkZyU1bvP8o1DYuNz5cejjaA_t7ks5qQPCHgaJpZM4I-HOt>
> .
>


-- 
-----
Best regards, Artem Polyakov
(Mobile mail)

**Comment by artpol84** [comment]
Again updated the PR https://github.com/hjelmn/ompi/pull/9 with slightly improved MPI_Waitsome performance according @bosilca [suggestion](https://github.com/open-mpi/ompi/pull/1820#issuecomment-228905750)

**Comment by jladd-mlnx** [comment]
@hjelmn @artpol84 @bosilca Do we have any idea why it's consistently hanging in init when using RDMACM? This is new behavior.

**Comment by artpol84** [comment]
I don't think it's related. The proble is in MPI Init. Unlikely request

вторник, 28 июня 2016 г. пользователь Joshua Ladd написал:

> @hjelmn <https://github.com/hjelmn> @artpol84
> <https://github.com/artpol84> @bosilca <https://github.com/bosilca> Do we
> have any idea why it's consistently hanging in init when using RDMACM? This
> is new behavior.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/open-mpi/ompi/pull/1816#issuecomment-229039372>, or mute
> the thread
> <https://github.com/notifications/unsubscribe/AHL5PiE6ZZUNF2sBKtWmNBxl5sU2cfgqks5qQRi7gaJpZM4I-HOt>
> .
>


-- 
-----
Best regards, Artem Polyakov
(Mobile mail)

**Comment by hjelmn** [comment]
Well, it shouldn't be crashing. It should be exiting. This is the problem:

```
06:44:23 --------------------------------------------------------------------------
06:44:23 No OpenFabrics connection schemes reported that they were able to be
06:44:23 used on a specific port.  As such, the openib BTL (OpenFabrics
06:44:23 support) will be disabled for this port.
06:44:23 
06:44:23   Local host:           jenkins01
06:44:23   Local device:         mlx5_0
06:44:23   Local port:           1
06:44:23   CPCs attempted:       rdmacm
06:44:23 --------------------------------------------------------------------------
```

rdmacm is working for me. Not sure why it is failing to load on Mellanox Jenkins.

**Comment by hjelmn** [comment]
Unless that is the other port.

**Comment by hjelmn** [comment]
BTW, if you need udcm to work with a router it wouldn't take much work. No need to use rdmacm for that :).

**Comment by artpol84** [comment]
I will check tomorrow.

**Comment by artpol84** [comment]
@hjelmn could you accept my PR?
@bosilca approved it.

**Comment by hjelmn** [comment]
@artpol84 Will merge after jenkins finishes.

**Comment by jsquyres** [comment]
@hjelmn @artpol84 We're less than 45 mins from the call -- do you want to wait for merging whichever PRs are going to get merged until we can do final coordination on the phone?

**Comment by artpol84** [comment]
Sure.
I was talking about my PR to @hjelmn branch. Let's talk.
